sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp

```diff
@@ -4,18 +4,635 @@
+          WINHTTP_NO_REQUEST_DATA,
+          0,
+          streamLength > 0 ? static_cast<DWORD>(streamLength) : 0,
+          0))
+  {
+    // Errors include:
+    // ERROR_WINHTTP_CANNOT_CONNECT
+    // ERROR_WINHTTP_CLIENT_AUTH_CERT_NEEDED
+    // ERROR_WINHTTP_CONNECTION_ERROR
+    // ERROR_WINHTTP_INCORRECT_HANDLE_STATE
+    // ERROR_WINHTTP_INCORRECT_HANDLE_TYPE
+    // ERROR_WINHTTP_INTERNAL_ERROR
+    // ERROR_WINHTTP_INVALID_URL
+    // ERROR_WINHTTP_LOGIN_FAILURE
+    // ERROR_WINHTTP_NAME_NOT_RESOLVED
+    // ERROR_WINHTTP_OPERATION_CANCELLED
+    // ERROR_WINHTTP_RESPONSE_DRAIN_OVERFLOW
+    // ERROR_WINHTTP_SECURE_FAILURE
+    // ERROR_WINHTTP_SHUTDOWN
+    // ERROR_WINHTTP_TIMEOUT
+    // ERROR_WINHTTP_UNRECOGNIZED_SCHEME
+    // ERROR_NOT_ENOUGH_MEMORY
+    // ERROR_INVALID_PARAMETER
+    // ERROR_WINHTTP_RESEND_REQUEST
+    GetErrorAndThrow("Error while sending a request.");
+  }
+
+  if (streamLength > 0 || streamLength == -1)
+  {
+    Upload(handleManager);
+  }
+}
+
+void WinHttpTransport::ReceiveResponse(std::unique_ptr<Details::HandleManager>& handleManager)
+{
+  // Wait to receive the response to the HTTP request initiated by WinHttpSendRequest.
+  // When WinHttpReceiveResponse completes successfully, the status code and response headers have
+  // been received.
+  if (!WinHttpReceiveResponse(handleManager->m_requestHandle, NULL))
+  {
+    // Errors include:
+    // ERROR_WINHTTP_CANNOT_CONNECT
+    // ERROR_WINHTTP_CHUNKED_ENCODING_HEADER_SIZE_OVERFLOW
+    // ERROR_WINHTTP_CLIENT_AUTH_CERT_NEEDED
+    // ...
+    // ERROR_WINHTTP_TIMEOUT
+    // ERROR_WINHTTP_UNRECOGNIZED_SCHEME
+    // ERROR_NOT_ENOUGH_MEMORY
+    GetErrorAndThrow("Error while receiving a response.");
+  }
+}
+
+int64_t WinHttpTransport::GetContentLength(
+    std::unique_ptr<Details::HandleManager>& handleManager,
+    HttpMethod requestMethod,
+    HttpStatusCode responseStatusCode)
+{
+  DWORD dwContentLength = 0;
+  DWORD dwSize = sizeof(dwContentLength);
+
+  // For Head request, set the length of body response to 0.
+  // Response will give us content-length as if we were not doing Head saying what would be the
+  // length of the body. However, server won't send any body.
+  // For NoContent status code, also need to set contentLength to 0.
+  int64_t contentLength = 0;
+
+  // Get the content length as a number.
+  if (requestMethod != HttpMethod::Head && responseStatusCode != HttpStatusCode::NoContent)
+  {
+    if (!WinHttpQueryHeaders(
+            handleManager->m_requestHandle,
+            WINHTTP_QUERY_CONTENT_LENGTH | WINHTTP_QUERY_FLAG_NUMBER,
+            WINHTTP_HEADER_NAME_BY_INDEX,
+            &dwContentLength,
+            &dwSize,
+            WINHTTP_NO_HEADER_INDEX))
+    {
+      contentLength = -1;
+    }
+    else
+    {
+      contentLength = static_cast<int64_t>(dwContentLength);
+    }
+  }
+
+  return contentLength;
+}
+
+std::unique_ptr<Details::WinHttpStream> WinHttpTransport::GetBodyStream(
```

Comment:
Does it matter? This approach is more flexible since the caller can up-cast as needed. Being specific at the method allows for that and being general loses the type info.